## 1. Importing Libraries

In [3]:
import os

import pandas as pd

import numpy as np

from sklearn.model_selection import train_test_split

In [4]:
!pip install pandas openpyxl

## 2. Read the data and cleaning starts

In [5]:
PROJECT_DIR = r"D:\Coding\AWS Flight Fare Prediction"
DATA_DIR = "Data"

In [6]:
def get_data(name):
    file_name = f"{name}.xlsx"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    return pd.read_excel(file_path)

In [7]:
flights = get_data("Flight_Fare_Prediction")

In [8]:
flights

Date_of_Journey   Airlines Airlines_code     Class Departure  \
0          2024-05-14     Indigo       6E-2519   Economy     21:26   
1          2024-10-22   GO FIRST        G8-354   Economy     04:52   
2          2024-10-21   SpiceJet       SG-8709   Economy     13:28   
3          2024-06-09  Air India        AI-805   Economy     08:00   
4          2024-05-31  Air India        AI-605   Economy     12:22   
...               ...        ...           ...       ...       ...   
10677      2024-03-10  Air India        AI-811  Business     13:52   
10678      2024-04-26  Air India        AI-479  Business     15:56   
10679      2024-07-05  Air India        AI-463  Business     17:36   
10680      2024-06-19  Air India        AI-471  Business     20:25   
10681      2024-09-30    Vistara        UK-635  Business     09:05   

      Total_Stops Arrival  Duration_in_Hours  Price_fare    Source  \
0        non-stop   11:16             2.3333        5899  Banglore   
1         2 stops   23:58             2.1667        5801   Kolkata   
2         2 stops   02:07             2.0833        5794     Delhi   
3          1 stop   02:23             2.1667        5955   Kolkata   
4          1 stop   21:15             2.2500        5955  Banglore   
...           ...     ...                ...         ...       ...   
10677    non-stop   05:03            21.7500       32646  Banglore   
10678    non-stop   09:39            20.6667       35532   Kolkata   
10679    non-stop   15:03            27.6667       35532   Kolkata   
10680    non-stop   01:04             5.3333       37256  Banglore   
10681    non-stop   17:45             4.8333       37417  Banglore   

      Destination                  Route  Unnamed: 12 Unnamed: 13  
0       New Delhi              BLR → DEL          NaN  2024-01-01  
1        Banglore  CCU → IXR → BBI → BLR          NaN  2024-12-31  
2          Cochin  DEL → LKO → BOM → COK          NaN         NaT  
3        Banglore        CCU → NAG → BLR          NaN         NaT  
4       New Delhi        BLR → NAG → DEL          NaN         NaT  
...           ...                    ...          ...         ...  
10677       Delhi              BLR → DEL          NaN         NaT  
10678    Banglore              CCU → BLR          NaN         NaT  
10679    Banglore              CCU → BLR          NaN         NaT  
10680       Delhi              BLR → DEL          NaN         NaT  
10681   New Delhi              BLR → DEL          NaN         NaT  

[10682 rows x 14 columns]

In [9]:
flights.drop('Unnamed: 12',axis=1,inplace=True)
flights.drop('Unnamed: 13',axis=1,inplace=True)

In [10]:
flights

Date_of_Journey   Airlines Airlines_code     Class Departure  \
0          2024-05-14     Indigo       6E-2519   Economy     21:26   
1          2024-10-22   GO FIRST        G8-354   Economy     04:52   
2          2024-10-21   SpiceJet       SG-8709   Economy     13:28   
3          2024-06-09  Air India        AI-805   Economy     08:00   
4          2024-05-31  Air India        AI-605   Economy     12:22   
...               ...        ...           ...       ...       ...   
10677      2024-03-10  Air India        AI-811  Business     13:52   
10678      2024-04-26  Air India        AI-479  Business     15:56   
10679      2024-07-05  Air India        AI-463  Business     17:36   
10680      2024-06-19  Air India        AI-471  Business     20:25   
10681      2024-09-30    Vistara        UK-635  Business     09:05   

      Total_Stops Arrival  Duration_in_Hours  Price_fare    Source  \
0        non-stop   11:16             2.3333        5899  Banglore   
1         2 stops   23:58             2.1667        5801   Kolkata   
2         2 stops   02:07             2.0833        5794     Delhi   
3          1 stop   02:23             2.1667        5955   Kolkata   
4          1 stop   21:15             2.2500        5955  Banglore   
...           ...     ...                ...         ...       ...   
10677    non-stop   05:03            21.7500       32646  Banglore   
10678    non-stop   09:39            20.6667       35532   Kolkata   
10679    non-stop   15:03            27.6667       35532   Kolkata   
10680    non-stop   01:04             5.3333       37256  Banglore   
10681    non-stop   17:45             4.8333       37417  Banglore   

      Destination                  Route  
0       New Delhi              BLR → DEL  
1        Banglore  CCU → IXR → BBI → BLR  
2          Cochin  DEL → LKO → BOM → COK  
3        Banglore        CCU → NAG → BLR  
4       New Delhi        BLR → NAG → DEL  
...           ...                    ...  
10677       Delhi              BLR → DEL  
10678    Banglore              CCU → BLR  
10679    Banglore              CCU → BLR  
10680       Delhi              BLR → DEL  
10681   New Delhi              BLR → DEL  

[10682 rows x 12 columns]

In [11]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10682 entries, 0 to 10681
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Date_of_Journey    10682 non-null  datetime64[ns]
 1   Airlines           10682 non-null  object        
 2   Airlines_code      10682 non-null  object        
 3   Class              10682 non-null  object        
 4   Departure          10682 non-null  object        
 5   Total_Stops        10681 non-null  object        
 6   Arrival            10682 non-null  object        
 7   Duration_in_Hours  10682 non-null  float64       
 8   Price_fare         10682 non-null  int64         
 9   Source             10682 non-null  object        
 10  Destination        10682 non-null  object        
 11  Route              10681 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(9)
memory usage: 1001.6+ KB


- We have 1 missing value in Total_Stops and Route Column
- We have total 10682 rows and 11 columns
- The data type of columns are inappropriate

## 3. Preliminary Analysis

### 3.1 Check Data Types

In [12]:
flights.dtypes

Date_of_Journey      datetime64[ns]
Airlines                     object
Airlines_code                object
Class                        object
Departure                    object
Total_Stops                  object
Arrival                      object
Duration_in_Hours           float64
Price_fare                    int64
Source                       object
Destination                  object
Route                        object
dtype: object

In [13]:
value = flights.Date_of_Journey.iloc[7]
value

Timestamp('2024-08-23 00:00:00')

In [14]:
value_1 = flights.Departure.iloc[2]
value_1

'13:28'

In [15]:
value_2 = flights.Arrival.iloc[2]
value_2

'02:07'

In [16]:
value_3 = flights.Duration_in_Hours.iloc[5]
value_3

2.25

In [17]:
value_4 = flights.Total_Stops.iloc[4]
value_4

'1 stop'

### 3.2 Check For Duplicates

In [18]:
flights.duplicated().sum()

0

In [19]:
(
    flights.
    loc[flights.duplicated(keep=False)]
    .sort_values(['Airlines','Date_of_Journey','Source','Destination'])
)
# Here loc function help us to filter the duplicate values more specifically 
# Keep = False means it counts all duplicate values and shows which values are same.

Empty DataFrame
Columns: [Date_of_Journey, Airlines, Airlines_code, Class, Departure, Total_Stops, Arrival, Duration_in_Hours, Price_fare, Source, Destination, Route]
Index: []

### 3.3 Observations
- The type of arrival and departure time should be datetime data type instead of string or object datatype.
- The total_stops has mixed data type but they should be numeric data type.
- There are 4 duplicate values in a dataset.

## 4. Detailed Analysis

In [20]:
flights.Airlines

0           Indigo
1         GO FIRST
2         SpiceJet
3        Air India
4        Air India
           ...    
10677    Air India
10678    Air India
10679    Air India
10680    Air India
10681      Vistara
Name: Airlines, Length: 10682, dtype: object

In [21]:
flights.Airlines.unique()

array(['Indigo', 'GO FIRST', 'SpiceJet', 'Air India', 'AirAsia',
       'Vistara', 'AkasaAir'], dtype=object)

- Here we have no inconsistence values in Airlines column

In [22]:
(
    flights
    .Airlines
    .str.replace("FIRST","First")
    .str.replace("Asia", " asia")
    .str.replace("AkasaAir", "Akasa air")
    .str.title() # Title method will capitalize the letter after the space
    .unique()
)

array(['Indigo', 'Go First', 'Spicejet', 'Air India', 'Air Asia',
       'Vistara', 'Akasa Air'], dtype=object)

### Source

In [23]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [24]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

### Departure

In [25]:
# Checking if it contains any colon instead of ":" in time 
(
    flights
    .Departure
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
)

Series([], Name: Departure, dtype: object)

In [26]:
(
    flights
    .Duration_in_Hours.round(2)
)

0         2.33
1         2.17
2         2.08
3         2.17
4         2.25
         ...  
10677    21.75
10678    20.67
10679    27.67
10680     5.33
10681     4.83
Name: Duration_in_Hours, Length: 10682, dtype: float64

### Total_Stops

In [27]:
(
    flights
    .Total_Stops
    .str.replace("non-stop","0")
    .str.replace(" stops?","",regex=True) # it will search for this pattern and ? either it will consider stop or stops
    .pipe(lambda ser: pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10677    0.0
10678    0.0
10679    0.0
10680    0.0
10681    0.0
Name: Total_Stops, Length: 10682, dtype: float64

## 5. Cleaning Operations

In [36]:
# ** This is the operator for unpacking the dictionary simply unpack the dictionary inside which we can access all the columns
# using dictionary comprehension concept.

from sklearn.impute import SimpleImputer
def clean_data(df):
    return (
        df
        .drop_duplicates()
        .assign(**{
            col: df[col].str.strip()
            for col in df.select_dtypes(include="O").columns
        })
        .rename(columns=str.lower)
        .assign (
            airlines = lambda df_: (
                df_
                .airlines
                .str.replace("FIRST","First")
                .str.replace("Asia", " asia")
                .str.replace("AkasaAir", "Akasa air")
                .str.title()
                
                
            ),
            departure = lambda df_: pd.to_datetime(df_.departure,format="mixed").dt.time,
            arrival = lambda df_: pd.to_datetime(df_.arrival,format="mixed").dt.time,
            duration_in_hours = lambda df_: df_.duration_in_hours.round(2)*60,
            total_stops = lambda df_: (
                df_
                .total_stops
                .str.replace("non-stop","0")
                .str.replace(" stops?","",regex=True) # it will search for this pattern and ? either it will consider stop or stops
                .pipe(lambda ser: pd.to_numeric(ser)) 
            )
        
            
        )
        .rename(columns = {'duration_in_hours':'duration_in_minutes'})
        .drop(columns="route")
        .ffill() # It replace the missing values with previous rows present in the dataset and it applies to entire rows which are null
    )

In [38]:
flights_data_cleaned = clean_data(flights)
flights_data_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10682 entries, 0 to 10681
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   date_of_journey      10682 non-null  datetime64[ns]
 1   airlines             10682 non-null  object        
 2   airlines_code        10682 non-null  object        
 3   class                10682 non-null  object        
 4   departure            10682 non-null  object        
 5   total_stops          10682 non-null  float64       
 6   arrival              10682 non-null  object        
 7   duration_in_minutes  10682 non-null  float64       
 8   price_fare           10682 non-null  int64         
 9   source               10682 non-null  object        
 10  destination          10682 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(7)
memory usage: 918.1+ KB


## 6. Split the Dataset in train test

In [39]:
X = flights_data_cleaned.drop(columns='price_fare')
y= flights_data_cleaned.price_fare.copy()

In [40]:
X_,X_test,y_,y_test = train_test_split(X,y,test_size=0.2,random_state=41)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=0.2,random_state=41)

print(X_train.shape,y_train.shape)
print(X_val.shape,y_val.shape)
print(X_test.shape,y_test.shape)

(6836, 10) (6836,)
(1709, 10) (1709,)
(2137, 10) (2137,)


## 7. Export the Subsets

In [41]:
def export_flights_data(X,y,name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)

    X.join(y).to_csv(file_path,index=False)

    return pd.read_csv(file_path).head()

In [42]:
export_flights_data(X_train,y_train,"train")

date_of_journey   airlines airlines_code     class departure  total_stops  \
0      2024-08-27    Vistara        UK-959  Business  12:07:00          1.0   
1      2024-05-10  Air India        AI-805   Economy  09:22:00          1.0   
2      2024-07-30   Spicejet       SG-8169   Economy  16:51:00          0.0   
3      2024-11-18  Air India        AI-481  Business  08:20:00          2.0   
4      2024-10-12    Vistara        UK-817   Economy  06:53:00          0.0   

    arrival  duration_in_minutes    source destination  price_fare  
0  21:52:00                735.0     Delhi      Cochin       37193  
1  23:39:00                130.2     Delhi      Cochin        7320  
2  04:30:00                124.8  Banglore       Delhi        5021  
3  04:43:00                814.8   Kolkata    Banglore       31462  
4  17:29:00                435.0  Banglore   New Delhi       10208

In [43]:
export_flights_data(X_test,y_test,"test")

date_of_journey   airlines airlines_code            class departure  \
0      2024-05-06    Vistara        UK-969  Premium Economy  12:25:00   
1      2024-05-01    Vistara        UK-993          Economy  17:25:00   
2      2024-12-22  Air India        AI-473         Business  17:51:00   
3      2024-06-19    Vistara        UK-879          Economy  12:29:00   
4      2024-03-30  Air India        AI-471          Economy  00:52:00   

   total_stops   arrival  duration_in_minutes    source destination  \
0          0.0  14:49:00                735.0  Banglore       Delhi   
1          0.0  00:45:00                130.2  Banglore       Delhi   
2          0.0  19:11:00               1035.0  Banglore   New Delhi   
3          2.0  07:32:00                880.2   Kolkata    Banglore   
4          1.0  04:43:00                289.8     Delhi      Cochin   

   price_fare  
0        7984  
1       13830  
2       39859  
3       13620  
4        8160

In [44]:
export_flights_data(X_val,y_val,"validation")

date_of_journey   airlines airlines_code            class departure  \
0      2024-07-28    Vistara        UK-706  Premium Economy  15:25:00   
1      2024-08-29  Air India        AI-512          Economy  12:47:00   
2      2024-01-15   Go First        G8-213          Economy  14:43:00   
3      2024-02-19    Vistara        UK-899         Business  01:40:00   
4      2024-06-26   Spicejet       SG-8709          Economy  22:29:00   

   total_stops   arrival  duration_in_minutes   source destination  price_fare  
0          2.0  01:41:00                495.0    Delhi      Cochin       18142  
1          1.0  19:46:00                964.8    Delhi      Cochin        7940  
2          1.0  02:20:00                505.2    Delhi      Cochin       17313  
3          1.0  04:21:00                505.2  Kolkata    Banglore       55696  
4          2.0  00:27:00                124.8    Delhi      Cochin        5794